packages/repository-json-schema/src/build-schema.ts

```diff
@@ -171,12 +175,15 @@ export function modelToJsonSchema(ctor: Function): JSONSchema {
     result.properties[p] = metaToJsonProperty(metaProperty);
 
     // populating JSON Schema 'definitions'
-    const referenceType = isArrayType(metaProperty.type as string | Function)
+    let referenceType = isArrayType(metaProperty.type as string | Function)
       ? // shimks: ugly type casting; this should be replaced by logic to throw
         // error if itemType/type is not a string or a function
         (metaProperty.itemType as string | Function)
       : (metaProperty.type as string | Function);
     if (typeof referenceType === 'function' && isComplexType(referenceType)) {
+      referenceType = isModelResolver(referenceType)
+        ? referenceType()
+        : referenceType;
```

Comment:
I'm seeing many places to check and resolve. It's be nice to have a common `resolveType` function.